File: api/routers/context.py
"""Context API Router - 上下文快照管理"""

from typing import Any

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel

from api.database import SessionLocal
from api.dependencies import get_current_user
from api.services.context_service import ContextService
from api.services.exceptions import PermissionDeniedError, ResourceNotFoundError

router = APIRouter()


class CreateSnapshotRequest(BaseModel):
    """创建快照请求"""

    session_id: str
    event_id: str
    context_data: dict[str, Any]


class SnapshotResponse(BaseModel):
    """快照响应"""

    context_capture_id: str
    session_id: str
    event_id: str
    context_data: dict[str, Any]
    created_at: str


class SnapshotListResponse(BaseModel):
    """快照列表响应"""

    snapshots: list[dict[str, Any]]
    total: int
    limit: int
    offset: int


@router.post(
    "",
    response_model=SnapshotResponse,
    status_code=status.HTTP_201_CREATED,
    summary="创建上下文快照",
)
async def create_snapshot(
    request: CreateSnapshotRequest, current_user: dict = Depends(get_current_user)
):
    """创建上下文快照"""
    try:
        service = ContextService(SessionLocal)
        result = service.create_snapshot(
            user_id=current_user["user_id"],
            session_id=request.session_id,
            event_id=request.event_id,
            context_data=request.context_data,
        )
        return result
    except PermissionDeniedError as e:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(e))
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=f"创建快照失败: {e!s}"
        )


@router.get("", response_model=SnapshotListResponse, summary="列出上下文快照")
async def list_snapshots(
    session_id: str | None = Query(None),
    limit: int = Query(50, ge=1, le=100),
    offset: int = Query(0, ge=0),
    current_user: dict = Depends(get_current_user),
):
    """列出上下文快照"""
    try:
        service = ContextService(SessionLocal)
        return service.list_snapshots(
            user_id=current_user["user_id"], session_id=session_id, limit=limit, offset=offset
        )
    except PermissionDeniedError as e:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(e))


@router.get("/{context_capture_id}", response_model=SnapshotResponse, summary="获取上下文快照")
async def get_snapshot(context_capture_id: str, current_user: dict = Depends(get_current_user)):
    """获取上下文快照"""
    try:
        service = ContextService(SessionLocal)
        return service.get_snapshot(
            context_capture_id=context_capture_id, user_id=current_user["user_id"]
        )
    except ResourceNotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))

File: rust/crates/runtime/src/service_handlers/context.rs
use astra_services::context::*;

use crate::AppState;
use astra_core::ErrorResponse;
use axum::{
    Json,
    extract::{Path, Query, State},
    http::{HeaderMap, StatusCode},
};

pub async fn create_snapshot_handler(
    State(state): State<AppState>,
    headers: HeaderMap,
    Json(request): Json<SnapshotCreateRequest>,
) -> Result<(StatusCode, Json<SnapshotResponse>), (StatusCode, Json<ErrorResponse>)> {
    let user = state.auth_service.current_user(&headers).await?;
    let snap = state
        .context_service
        .create_snapshot(
            user.user_id,
            SnapshotCreateRequestData {
                session_id: request.session_id,
                event_id: request.event_id,
                context_data: request.context_data,
            },
        )
        .await?;
    Ok((StatusCode::CREATED, Json(SnapshotResponse::from(snap))))
}

pub async fn list_snapshots_handler(
    State(state): State<AppState>,
    headers: HeaderMap,
    Query(q): Query<SnapshotListQuery>,
) -> Result<Json<SnapshotListResponse>, (StatusCode, Json<ErrorResponse>)> {
    let user = state.auth_service.current_user(&headers).await?;
    let list = state
        .context_service
        .list_snapshots(SnapshotListFilter {
            user_id: user.user_id,
            session_id: q.session_id,
            limit: q.limit,
            offset: q.offset,
        })
        .await?;
    Ok(Json(SnapshotListResponse::from(list)))
}

pub async fn get_snapshot_handler(
    State(state): State<AppState>,
    Path(context_capture_id): Path<String>,
    headers: HeaderMap,
) -> Result<Json<SnapshotResponse>, (StatusCode, Json<ErrorResponse>)> {
    let user = state.auth_service.current_user(&headers).await?;
    let snap = state
        .context_service
        .get_snapshot(context_capture_id, user.user_id)
        .await?;
    Ok(Json(SnapshotResponse::from(snap)))
}
